xgboost_ray/util.py

```diff
@@ -113,3 +114,77 @@ def shutdown(self):
             if self.actor:
                 ray.kill(self.actor)
             self.actor = None
+
+
+class MultiActorTask:
+    """Utility class to hold multiple futures.
+
+    The `is_ready()` method will return True once all futures are ready.
+
+    Args:
+        pending_futures (list): List of object references (futures)
+            that should be tracked.
+    """
+
+    def __init__(self, pending_futures: Optional[List[ray.ObjectRef]] = None):
+        self._pending_futures = pending_futures or []
+        self._ready_futures = []
+
+    def is_ready(self):
+        if not self._pending_futures:
+            return True
+
+        ready = True
+        while ready:
+            ready, not_ready = ray.wait(self._pending_futures, timeout=0)
+            if ready:
+                for obj in ready:
+                    self._pending_futures.remove(obj)
+                    self._ready_futures.append(obj)
+
+        return not bool(self._pending_futures)
+
+
+def _num_possible_actors(num_cpus_per_actor: int,
+                         num_gpus_per_actor: int,
+                         resources_per_actor: Optional[Dict] = None,
+                         max_needed: int = -1) -> int:
+    """Returns number of actors that could be scheduled on this cluster."""
+    if max_needed < 0:
+        max_needed = float("inf")
+
+    resources_per_actor = resources_per_actor or {}
+
+    def _check_resources(resource_dict: Dict):
+        # Check how many actors could be scheduled given available
+        # resources in `resource_dict`
+        available_cpus = resource_dict.get("CPU", 0.)
+        available_gpus = resource_dict.get("GPU", 0.)
+        available_custom = {
+            k: resource_dict.get(k, 0.)
+            for k in resources_per_actor
+        }
+
+        actors_cpu = actors_gpu = actors_custom = float("inf")
+
+        if num_cpus_per_actor > 0:
+            actors_cpu = math.floor(available_cpus / num_cpus_per_actor)
+        if num_gpus_per_actor > 0:
+            actors_gpu = math.floor(available_gpus / num_gpus_per_actor)
+        if resources_per_actor:
+            actors_custom = min(
+                math.floor(available_custom[k] / resources_per_actor[k])
+                if available_custom[k] > 0. else float("inf")
+                for k in resources_per_actor)
+
+        return min(actors_cpu, actors_gpu, actors_custom)
+
+    num_possible_actors = 0
+    ray.state.state._check_connected()  # Initialize global state
+    for resources in ray.state.state._available_resources_per_node().values():
```

Comment:
This uses internal methods of Ray's global state as there is no public API to get available resources by node. Can we keep this or should I upstream a public API to ray core?

Reply:
Global cluster resources are not enough to check because we can't know how they are distributed across machines and thus don't know if scheduling an actor is actually feasible.

Reply:
instead of using `_check_connected` maybe instead use ray.worker.global_worker.check_connected()?

Reply:
_available_resources_per_node is used by serve too -- we should make it public! (ok to merge as is)

Reply:
Turns out we need `_check_connected` as it lazily initializes the global state accessor on latest master